### ISSUE TITLE
add confidence intervals to scipy.signal.welch function

### ISSUE THREAD

**Issue by jerabaul29** [scipy/scipy#6363]
Would it be possible to add the computations of confidence intervals to the welch function, for example with the same interface as provided with Matlab pwelch?

While this is not difficult to implement 'by hand' and some people provide help for this on the internet:

https://rhenanbartels.wordpress.com/2014/04/06/first-post/

, I think this would be a nice help for people who are not too familiar with probability and distributions.

**Comment by rgommers** [comment]
Makes sense to add. Would be good to be able to return the extra outputs unconditionally though, so this needs a ``Bunch`` class as being discussed right now at gh-6335

**Comment by jerabaul29** [comment]
It seems it do will be necessary to be very careful about how to compute the error bars. So far I get three different sources, and three different formula:

- (already cited in a previous post): https://rhenanbartels.wordpress.com/2014/04/06/first-post/
This should be if I understand well from a PhD guy, so probably should be reliable even if it is not an article

- http://www.osti.gov/scitech/servlets/purl/5688766
pages 29 (nu estimation) and 30. This is a Sandia National Laboratories report, so I would expect it to be reliable

- http://www.ndbc.noaa.gov/wavemeas.pdf
page 7, this is a technical notice from the US National Oceanic and Atmospheric Administration, so I would guess it should be reliable also...

I must confess I did not have the time to go through the technical details in each paper for checking if they use the exact same Welch estimate and so on. For the time being I will stuck to Matlab to avoid headaches about those issues. But if somebody develops a confidence interval function for the Welch method, can you document carefully what formula you use and validate it against Matlab?

**Comment by rgommers** [comment]
Thanks for the details @jerabaul29. Guess we'll leave this issue open as an enhancement request then, and see if someone interested picks up on it.

**Comment by e-q** [comment]
Given all the choices that can be made on how to calculate the confidence intervals, I think it may be better for users to calculate them from the output of `spectrogram` or `stft` themselves.

**Comment by jerabaul29** [comment]
Any update on this?

**Comment by ilayn** [comment]
Judging from the previous responses, it looks like nobody picked it up and it is waiting for volunteers.

**Comment by luhan93** [comment]
I agree that add the confidence interval is necessary, as my advisor always says "the number makes no sense without an error bar". I'd like to give it a try. I guess the simplest way is to replicate what Matlab does in pwelch.

**Comment by jerabaul29** [comment]
I think to maximally reduce risks of making mistakes (using the wrong number of segments, wrong formula depending on window and or overlap, etc), that it would be very useful to provide these natively out of the output - and / or at least, have a way to pass out the number of degrees of freedom in the chi-square law?

I am not a statistician so would be scary for me to try to write a PR for this, too many ways to shoot myself in the feet. But if this can be useful, I recently wrote down a small summary for my own use and clarifying my mind:

https://github.com/jerabaul29/tutorials/blob/main/uncertainty_gaussian_spectrum/uncertainty_gaussian_spectrum.ipynb

I usually use Hann with 50% overlap, so this is enough for my use, but probably something better and more general would be useful to the community.
